source/common/grpc/async_client_impl.h

```diff
@@ -0,0 +1,206 @@
+      return nullptr;
+    }
+
+    grpc_stream->set_stream(http_stream);
+
+    Http::MessagePtr message = Common::prepareHeaders(
+        remote_cluster_name_, service_method.service()->full_name(), service_method.name());
+    callbacks.onCreateInitialMetadata(message->headers());
+
+    http_stream->sendHeaders(message->headers(), false);
+    grpc_stream->moveIntoList(std::move(grpc_stream), active_streams_);
+
+    return active_streams_.front().get();
+  }
+
+private:
+  Upstream::ClusterManager& cm_;
+  const std::string remote_cluster_name_;
+  std::list<std::unique_ptr<AsyncClientStreamImpl<RequestType, ResponseType>>> active_streams_;
+
+  friend class AsyncClientStreamImpl<RequestType, ResponseType>;
+};
+
+template <class RequestType, class ResponseType>
+class AsyncClientStreamImpl : public AsyncClientStream<RequestType>,
+                              Http::AsyncClient::StreamCallbacks,
+                              LinkedObject<AsyncClientStreamImpl<RequestType, ResponseType>> {
+public:
+  AsyncClientStreamImpl(AsyncClientImpl<RequestType, ResponseType>& parent,
+                        AsyncClientCallbacks<ResponseType>& callbacks)
+      : parent_(parent), callbacks_(callbacks) {}
+
+  // Http::AsyncClient::StreamCallbacks
+  void onHeaders(Http::HeaderMapPtr&& headers, bool end_stream) override {
+    ASSERT(!remote_closed_);
+    const auto http_response_status = Http::Utility::getResponseStatus(*headers);
+    if (http_response_status != enumToInt(Http::Code::OK)) {
+      // https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md requires that
+      // grpc-status be used if available.
+      if (end_stream && Common::getGrpcStatus(*headers).valid()) {
+        onTrailers(std::move(headers));
+        return;
+      }
+      streamError(Common::httpToGrpcStatus(http_response_status));
+      return;
+    }
+    if (end_stream) {
+      onTrailers(std::move(headers));
+      return;
+    }
+    callbacks_.onReceiveInitialMetadata(std::move(headers));
+  }
+
+  void onData(Buffer::Instance& data, bool end_stream) override {
+    ASSERT(!remote_closed_);
+    std::vector<Frame> frames;
+    if (!decoder_.decode(data, frames)) {
+      streamError(Status::GrpcStatus::Internal);
+      return;
+    }
+
+    for (const auto& frame : frames) {
+      std::unique_ptr<ResponseType> response(new ResponseType());
+      // TODO(htuch): We can avoid linearizing the buffer here when Buffer::Instance implements
+      // protobuf ZeroCopyInputStream.
+      // TODO(htuch): Need to add support for compressed responses as well here.
+      if (frame.flags_ != GRPC_FH_DEFAULT ||
+          !response->ParseFromArray(frame.data_->linearize(frame.data_->length()),
+                                    frame.data_->length())) {
+        streamError(Status::GrpcStatus::Internal);
+        return;
+      }
+      callbacks_.onReceiveMessage(std::move(response));
+    }
+    if (end_stream) {
```

Comment:
If this is a protocol failure, should we check this before even trying to do the decode up above?